### ISSUE TITLE
Define the use of robotic station

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#1165]
**Describe the Issue**
- Currently, the robotic station function mimics life support function. 
- The robotic station function is for robots while life support is for human.
- However, most (if not all) robotic station function has activity spots while life support has none in buildings.xml.

**Questions to be resolved **
- What's the real use of robotic station ? For charging ? For parking ?
- Is it required for a robot to step into a building without robotic station ? 
- Is adding to a robotic station supposed to be the same as occupying an activity spot ?
- If a robot occupies an activity spot in a robotic station, what is it allowed to do ?
- How will a rover support hosting a robot in future ? 
- How will a robot work outside or in a off site remote location for future construction project ?

**Discuss any potential problems**
- When a robot changes a building, it will check if the desired function has an empty activity spot, but also an empty robotic station activity spot. This is duplicated checking.
- If a spot is available, then it will be transferred. 
- If this is the norm, then it does not require checking if the destination building has any empty robotic station activity spot.

**Java classes involved**
- `RoboticStation`

**Specifications**
- pre 3.7.0 build 9046

**Comment by mokun** [comment]
@bevans2000 

Can we give a definitive usage of a robotic station here ?

There are 3 roles:
1. being a part of robotOccupants 
2. occupying an activity spot
3. charging the battery

There's a Collection called robotOccupants in `RoboticStation` class : 

```
	private Collection<Robot> robotOccupants;
```

Currently, a robot being in this collection doesn't mean it's charging and doesn't mean it occupies an activity spot.

There are methods in `BuildingManager` that confuses those 3 roles above.

**Comment by mokun** [comment]
But we could imagine wireless charging. I actually prefer this route.

This frees up the need for a robot to walk back and forth between locations just to be able to charge the battery.

**Comment by mokun** [comment]
@bevans2000 

Now, if a robot somehow cannot find an open robotic station with an open activity spot, it can still be charging wirelessly, albeit at a slower rate.

**Comment by bevans2000** [comment]
Ok. I will integrate this change with my work on #1166

**Comment by bevans2000** [comment]
What else is to do on this issue?

**Comment by mokun** [comment]
This is from a SpaceX phase 1 settlement. I'm still getting *Depleted* battery from time to time. 

![image](https://github.com/mars-sim/mars-sim/assets/1168584/a0cbfc4b-e773-463d-acbe-d30042e9ad5f)

```
01-Adir-05:017.514 (Warning) SystemCondition [x256] : [Teslapolis] MakerBot 002 - Out of power.
01-Adir-05:061.574 (Warning) BotMind [x1380] : [Teslapolis] MakerBot 001 - Battery almost depleted and must be recharged. Current task: None.
...
01-Adir-05:217.889 (Warning) SystemCondition [x272] : [Teslapolis] MakerBot 001 - Out of power.
...
01-Adir-05:261.674 (Warning) SystemCondition [x262] : [Teslapolis] GardenBot 001 - Out of power.

```

**Comment by mokun** [comment]
Let's make sure no bots are sitting duck with a depleted battery. It shouldn't happen. 

Do you still see it happening ?

**Comment by mokun** [comment]
@bevans2000 

It turns out battery can still be depleted in 3.7.0. See below : 

![image](https://github.com/mars-sim/mars-sim/assets/1168584/34a22101-2a26-4540-abfe-6979f12fc09b)

**Comment by mokun** [comment]
I can now see why the spurious msg but with no action. Fixing it now on `3.7.x` branch.

**Comment by mokun** [comment]
Finally found out why it doesn't work. 

It has to do with not being able to walk successfully to a robotic station and claim the spot. 

```
				// Claim this activity spot
				boolean canClaim = f.claimActivitySpot(loc, person);
				
				if (!canClaim)
					result = false;
				else
					result = true;
```
I'll report it in https://github.com/mars-sim/mars-sim/issues/1166 for the record.

![image](https://github.com/mars-sim/mars-sim/assets/1168584/f76b5e8e-0add-4a09-a580-3ecaf3485fec)

![image](https://github.com/mars-sim/mars-sim/assets/1168584/2a10e584-2eb7-47bf-bda8-fa549e9883ab)

**Comment by mokun** [comment]
@bevans2000,

How do we make all tasks call certain methods ONLY ONCE and branch off into doing different thing, without having to  repeatedly carry out outdated actions ?

I know each situation is different.

So, based upon the log statement below, let's examine each specific case for these bots to see if the log statement reveal what logics we need to improve. Shall we ? 

```
01-Adir-07:979.160 (Warning) BotMind [x507] : [New Maktoum] RepairBot 003 - Battery almost depleted and must be recharged. Current task: Maintaining Building.
01-Adir-07:982.579 (Warning) SystemCondition [x1014] : [New Maktoum] RepairBot 001 - Out of power.
01-Adir-07:986.849 (Info) BotTaskManager [x339] : [New Maktoum] GardenBot 002 - Charging is triggered due to low power.
01-Adir-07:988.555 (Warning) BotMind [x1014] : [New Maktoum] GardenBot 001 - Battery almost depleted and must be recharged. Current task: Planning to recharge.
01-Adir-07:992.305 (Warning) SystemCondition [x508] : [New Maktoum] GardenBot 001 - Out of power.
01-Adir-07:994.009 (Warning) SystemCondition [x1016] : [New Maktoum] RepairBot 002 - Out of power.
```

Let's look at these 3 places first : 
1. BotMind - repeated 500 to 1000 times
2. BotTaskManager - repeated over 300 times
3. SystemCondition - repeated 500 to 1000 times

**Comment by bevans2000** [comment]
Ok, yes this looks odd. Let me finish #1072

**Comment by mokun** [comment]
@bevans2000 

Let's keep track of the battery depletion problem in this thread.

**Comment by mokun** [comment]
@bevans2000 

Here's the battery status on sol 2 : 

![image](https://github.com/mars-sim/mars-sim/assets/1168584/377687c9-6494-426c-9c85-98b0faf4e3ec)

It gets worse. 

When I come back 10 mins later on sol 7, 3 out of 4 bots are dead in the water.

![image](https://github.com/mars-sim/mars-sim/assets/1168584/bddcbe4a-67c1-490c-a89a-023e1427c06f)

Their tasks were "Walking to a robotic station" 

So they must have trouble finding an open robotic station somehow. But why ? This settlement has 4 buildings and it doesn't lack any robotic stations.

![image](https://github.com/mars-sim/mars-sim/assets/1168584/9900452e-5190-405b-b9c1-db5750d3bbd7)

**Comment by mokun** [comment]
So it's another classical "walking bug".

They all got stuck at try "walking to a robotic station". 

They never arrive anywhere when power is depleted.

They could have done "Wireless Charging" instead, albeit it's slower. 

But they got stuck at walking nowhere.

![image](https://github.com/mars-sim/mars-sim/assets/1168584/82442574-6385-461d-9ea9-b32c9a120363)

**Comment by mokun** [comment]
There are 8 robotic stations (charge ports) in this settlement.

![image](https://github.com/mars-sim/mars-sim/assets/1168584/a340a52e-680b-4e9a-a797-71546eb83001)

**Comment by bevans2000** [comment]
Yes it does look like the Robots are stuck. Have you debugged it?

**Comment by mokun** [comment]
> Yes it does look like the Robots are stuck. Have you debugged it?

I just cleared those up now. Committing.

The culprit is that it returns the same time without reducing it in Charge task.

So, in general, it's very dangerous to return the same amount of time in any phase method of a task.

**Comment by mokun** [comment]
@bevans2000 @Urwumpe 

In near future, we do need a separate building as an engineer's shop with lab benches for taking apart robots and doing maintenance on their parts.

To do a building maintenance, a person walks to that building to do it.

But for robot maintenance, we don't have a (meta) task suited for this purpose at this moment.

We can either have a robot walk to the shop in order to check itself in, or if it's malfunctioned and unable to walk, a person may haul it there on a robotic trolley.

**Comment by bevans2000** [comment]
Yes we can do that. Makes sense.
But I would just use the existing Manufacturing Function because that represents a workshop and tools. Maybe add a new Tools type for robot

**Comment by Urwumpe** [comment]
> [@bevans2000](https://github.com/bevans2000) [@Urwumpe](https://github.com/Urwumpe)
> 
> In near future, we do need a separate building as an engineer's shop with lab benches for taking apart robots and doing maintenance on their parts.

Still, I would prefer some scalability in that matter. Maybe a long term goal, but I think it would be better, if specialization of workplaces:

-  can come naturally with the growth of the settlement
- allows multiple different approaches depending on settlement priorities or available resources
- can have different levels of productivity and fixed efforts.

In the MVP sense, the smallest viable robotic workplace should be just a parked robot (If maintenance could be done just as EVA or IVA). Of course, not as capable (again, think dust or missing tools/spare parts) and more time consuming for one astronaut.  And everything else develop from there. 

If a base has the infrastructure to support a huge highly automated assembly line for robots and can satisfy the needs of other bases for robots and spare parts - why not? It would mean that the settlers in general have more time for other tasks or that less resupply capacity must be allocated to robots and spareparts. Same if the settlers could install recharge and cleaning stations for their robots.

**Comment by bevans2000** [comment]
We need to adopt this pattern in ```Building``` to remove the code bloat around the getters.
```
  private Map<FunctionType, ? extends Function> functions;

  public <T extends Function> T getFunction(FunctionType key) {
          return (T) functions.get(key);
   }
```
This will need a lot of method change in the callers but break the coupling between the Functions and the Building. Would need a FunctionFactory class as well. We need to plan when to do this

**Comment by mokun** [comment]
>  applies of course to the workplaces used there. A robot could be repaired in the hab on the kitchen desk. Or at least parked somewhere sheltered. 

That's true. Initially, all they need is a space for a repair shop. 

Over time, it will grow and differentiate which shop is good for what purpose.

You can say this is the flaw in all strategy games. The player dictates what to build and often times there are surplus of a type of buildings. 

And there's no way to retrofit that building into something else.

So now the question is, how do we model to have a generic workspace or a building ? 

How do we allow the simulation or the settlement to decide what it will need ?

It will be very interesting to allow this emergent self-organizational behavior.

I'm up for building an evolutionary process that let the settlement reconfigure and convert a building into a new workspace that meet its current objectives and needs.

**Comment by mokun** [comment]
@bevans2000 

```java
  private Map<FunctionType, ? extends Function> functions;

  public <T extends Function> T getFunction(FunctionType key) {
          return (T) functions.get(key);
   }
```

> would have to use a EnumMap but that lookup will be super quick

This morning I'm working on replacing the functions list in Building with enum map. 

However, I run into a road block.  

```java
	/** A list of functions of this building. */
//	private List<Function> functions = new ArrayList<>();
	
	/** A map of functions for this building. */
	private Map<FunctionType, ? extends Function> functions = new EnumMap<>(FunctionType.class);
	
	/**
	 * Determines the building functions.
	 *
	 * @return list of building .
	 * @throws Exception if error in functions.
	 */
	public Function addFunction(FunctionSpec fSpec) {
		var f =  switch (fSpec.getType()) {
			case ADMINISTRATION -> new Administration(this, fSpec);
			case ALGAE_FARMING -> new AlgaeFarming(this, fSpec);				
			case ASTRONOMICAL_OBSERVATION -> new AstronomicalObservation(this, fSpec);
			case CONNECTION -> new BuildingConnection(this, fSpec);
			case COMMUNICATION -> new Communication(this, fSpec);
			case COMPUTATION -> new Computation(this, fSpec);
			case COOKING -> new Cooking(this, fSpec);
			case DINING -> new Dining(this, fSpec);
			case EARTH_RETURN -> new EarthReturn(this, fSpec);
			case EVA -> new EVA(this, fSpec);
			case EXERCISE -> new Exercise(this, fSpec);
			case FARMING -> new Farming(this, fSpec);
			case FISHERY -> new Fishery(this, fSpec);
			case FOOD_PRODUCTION -> new FoodProduction(this, fSpec);
			case VEHICLE_MAINTENANCE -> new VehicleGarage(this, fSpec);
			case LIFE_SUPPORT -> new LifeSupport(this, fSpec);
			case LIVING_ACCOMMODATION -> new LivingAccommodation(this, fSpec);
			case MANAGEMENT -> new Management(this, fSpec);
			case MANUFACTURE -> new Manufacture(this, fSpec);
			case MEDICAL_CARE -> new MedicalCare(this, fSpec);
			case POWER_GENERATION -> new PowerGeneration(this, fSpec);
			case POWER_STORAGE -> new PowerStorage(this, fSpec);
			case RECREATION -> new Recreation(this, fSpec);
			case RESEARCH -> new Research(this, fSpec);
			case RESOURCE_PROCESSING -> new ResourceProcessing(this, fSpec);
			case ROBOTIC_STATION -> new RoboticStation(this, fSpec);
			case STORAGE -> new Storage(this, fSpec);
			case THERMAL_GENERATION -> new ThermalGeneration(this, fSpec);
			case WASTE_PROCESSING -> new WasteProcessing(this, fSpec);
			default ->
				throw new IllegalArgumentException("Do not know how to build Function " + fSpec.getType());
			};

		// Add to the function map
		functions.put(fSpec.getType(), f);
		
		return f;
	}
```

It won't let me use this 
`functions.put(fSpec.getType(), f);`

Eclipse IDE said 
`The method put(FunctionType, capture#4-of ? extends Function) in the type Map<FunctionType,capture#4-of ? extends Function> is not applicable for the arguments (FunctionType, Function)`

How do I fix this ?

**Comment by bevans2000** [comment]
Hmm, ok. If it works that's fine

**Comment by bevans2000** [comment]
> For now, many methods below in `RoboticStation` are being called up.
> 
> There's no sure way to replace `private Collection<Robot> robotOccupants` yet.
> 
Maybe we just hold a list of Workers against the Building and consolidate it with the Person list in the Life support function?

This would support the case where a Person was inside and outside building that didn't have Life Support. At the moment we can not model that.

Could be a breaking change but it would be an improvement.

**Comment by mokun** [comment]
> This would support the case where a Person was inside and outside building that didn't have Life Support. 

Can you explain more ?

**Comment by bevans2000** [comment]
Nice, but do you need the cast in the specific get methods? The next question is why have the specialists get method at all? Dropping the would cut down on the boilerplate code in Building
